### ISSUE TITLE
Uninformative error message for no permissions to /var/lib/scylla/

### ISSUE THREAD

**Issue by noamha** [scylladb/scylla#883]
Configuration: 1 x Rackspace OnMetal IO v1, Scylla 0.17 build 10

I played around with /var/lib/scylla, and in one of the cases I created a new mount which had root as owner
When trying to start the service I got the following error from the bash:

```
Redirecting to /bin/systemctl start  scylla-server.service
Job for scylla-server.service failed. See 'systemctl status scylla-server.service' and 'journalctl -xn' for details.
```

The journalctl logs showed:

```
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 31 as core 1 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 32 as core 2 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 33 as core 3 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 34 as core 4 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 35 as core 8 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 36 as core 9 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 37 as core 10 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 38 as core 11 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected lcore 39 as core 12 on socket 1
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Support maximum 128 logical core(s) by configuration.
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: EAL: Detected 40 lcore(s)
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: PANIC in rte_eal_config_create():
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: Cannot open '/var/lib/scylla/.rte_config' for rte_mem_config
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 9: [/usr/bin/scylla() [0x4c64e1]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 8: [/lib64/libc.so.6(__libc_start_main+0xf5) [0x7f747acaaaf5]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 7: [/usr/bin/scylla() [0x4703ab]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 6: [/usr/bin/scylla() [0x569eac]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 5: [/usr/bin/scylla() [0x51744f]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 4: [/usr/bin/scylla() [0x570dfb]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 3: [/usr/bin/scylla() [0x6bf371]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 2: [/usr/bin/scylla() [0x469f3a]]
Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: 1: [/usr/bin/scylla() [0x6c5328]]
Feb 08 13:58:29 noam-server-01.localdomain systemd-coredump[38852]: Coredump too large, ignoring
Feb 08 13:58:29 noam-server-01.localdomain systemd-coredump[38852]: Process 38851 (scylla) dumped core.
Feb 08 13:58:29 noam-server-01.localdomain systemd[1]: scylla-server.service: main process exited, code=dumped, status=6/ABRT
Feb 08 13:58:29 noam-server-01.localdomain sudo[38854]: scylla : TTY=unknown ; PWD=/var/lib/scylla ; USER=root ; COMMAND=/usr/lib/scylla/scylla_stop
Feb 08 13:58:29 noam-server-01.localdomain systemd[1]: Failed to start Scylla Server.
```

Instead of a clear error message regarding the permission issue for the folder

**Comment by raphaelsc** [comment]
@noamha, dorlaor: DPDK is responsible for opening that config file and showing the error message. What are we doing when we aren't happy with some DPDK code? Are we patching it? Are we trying to push the change mainstream?

**Comment by vladzcloudius** [comment]
On Feb 10, 2016 18:50, "Raphael Carvalho" <notifications@github.com> wrote:
>
> @noamha, dorlaor: DPDK is responsible for opening that config file and
showing the error message. What are we doing when we aren't happy with some
DPDK code? Are we patching it? Are we trying to push the change mainstream?

@raphael we are patching it and pushing upstream but it doesn't look like a
DPDK problem - seems like it's the same permissions problem we were
discussing on a different thread.

**Comment by raphaelsc** [comment]
@vladzcloudius Got it, thanks. But if I understand @noamha correctly, he was only worried about the error message (printed by rte_eal_config_create()) not being clear enough.

**Comment by vladzcloudius** [comment]
On 02/10/2016 07:18 PM, Raphael Carvalho wrote:
>
> @vladzcloudius <https://github.com/vladzcloudius> Got it, thanks. But 
> if I understand @noamha <https://github.com/noamha> correctly, he was 
> only worried about the error message (printed by 
> rte_eal_config_create()) not being clear enough.
>

I see. IMHO the message was crystal clear - there is a problem with the 
permissions (probably). And this is what the

|Cannot open '/var/lib/scylla/.rte_config' for rte_mem_config |

||is all about. How clearer it could be?

**Comment by nyh** [comment]
@vladzcloudius, am I correct in understanding that this ".rte_config" is ***not*** a configuration file which DPDK needs to read, but rather - a file which DPDK tries do *write*?

First of all, I think it absolutely stupid it uses files at all for this purpose (it should keep this information in memory). 

Second, yes, @vladzcloudius, the messages could have been clearer - instead of "cannot open" it should have said cannot write, or something of this sort, and avoid this phrase "for rte_mem_config" (who the heck knows what rte_mem_config is???)

Third, most of the messages we see in this log should not have appeared, and the real problem would have stood out more easily. Why do we need to see 40 lines, one for each core detected by DPDK? When we're not even using DPDK? :-)

Finally, there shouldn't have been a core dump attempt at the end. This is an error and should have caused a clean exit - it's not a bug which should cause a core dump.

I think these most of the above need to be fixed in DPDK, not scylla...

**Comment by nyh** [comment]
Oh, and of course, I'll ask again a question I asked already a dozen times before: Do we ***really*** need to start DPDK even when it isn't used in this run? I know coding it this way simplified a few things, but maybe it's time to go back and change this part. The reason I'm asking for this is to have fewer chances of DPDK-related bugs which crop up in our non-DPDK runs. These bugs are not hypothetical - before this one, we recently had a signal-related bug which Avi tracked to DPDK's extra threads.

**Comment by vladzcloudius** [comment]
On 02/10/2016 11:31 PM, nyh wrote:
>
> @vladzcloudius <https://github.com/vladzcloudius>, am I correct in 
> understanding that this ".rte_config" is */not/* a configuration file 
> which DPDK needs to read, but rather - a file which DPDK tries do /write/?
>
> First of all, I think it absolutely stupid it uses files at all for 
> this purpose (it should keep this information in memory).
>

If memory serves me well DPDK**is also using this file to prevent two 
/master/ DPDK processes from running at the same time. So using a FS is 
intentional here.

> Second, yes, @vladzcloudius <https://github.com/vladzcloudius>, the 
> messages could have been clearer - instead of "cannot open" it should 
> have said cannot write, or something of this sort, and avoid this 
> phrase "for rte_mem_config"
>

If they fail to even open it why would they report that they failed to 
write into it?

> (who the heck knows what rte_mem_config is???)
>

We (ScyllaDB Ltd.) should know. DPDK is not intended to be used directly 
by a "non-aware" users and the messages above are meant to be read by 
the well-aware personnel. It's up to us to prevent them from being 
printed if don't need them anymore. I don't think it's fair to blame 
DPDK for then thing we haven't done...

> Third, most of the messages we see in this log should not have 
> appeared, and the real problem would have stood out more easily. Why 
> do we need to see 40 lines, one for each core detected by DPDK? When 
> we're not even using DPDK? :-)
>

We do use DPDK. If u see these messages means we use DPDK. In particular 
we use its sharding capability - our seastar's per-shard "engine" is 
running inside DPDK's thread.

> Finally, there shouldn't have been a core dump attempt at the end. 
> This is an error and should have caused a clean exit - it's not a bug 
> which should cause a core dump.
>
> I think these most of the above need to be fixed in DPDK, not scylla...
>

DPDK maybe calls for rte_panic() in too many places during the 
initialization - that's true. We can either cook a patch fixing this 
specific place (and this would not be a single line patch) or we may go 
and make sure in our installation scripts that the permissions are as 
expected. The second would be MUCH easier.
Or alternatively - we may compile the DPDK out as u suggested and then 
we will have to maintain two distributions - one with DPDK support and 
another without and IMHO this is not justified.

**Comment by vladzcloudius** [comment]
On 02/10/2016 11:35 PM, nyh wrote:
>
> Oh, and of course, I'll ask again a question I asked already a dozen 
> times before: Do we */really/* need to start DPDK even when it isn't 
> used in this run? I know coding it this way simplified a few things, 
> but maybe it's time to go back and change this part. The reason I'm 
> asking for this is to have fewer chances of DPDK-related bugs which 
> crop up in our non-DPDK runs. These bugs are not hypothetical - before 
> this one, we recently had a signal-related bug which Avi tracked to 
> DPDK's extra threads.
>

@nyh, I think the main reason we compile DPDK in even if we do not 
intend to use its networking backend right now is to be able to maintain 
a single binary when we *will *prefer**/native/ seastar's networking 
backend. I hope this is going to be quite soon right after GA.

**Comment by nyh** [comment]
On Thu, Feb 11, 2016 at 1:04 PM, vladzcloudius <notifications@github.com>
wrote:

>
> On 02/10/2016 11:31 PM, nyh wrote:
> >
> > @vladzcloudius <https://github.com/vladzcloudius>, am I correct in
> > understanding that this ".rte_config" is */not/* a configuration file
> > which DPDK needs to read, but rather - a file which DPDK tries do
> /write/?
> >
> > First of all, I think it absolutely stupid it uses files at all for
> > this purpose (it should keep this information in memory).
> >
>
> If memory serves me well DPDK**is also using this file to prevent two
> /master/ DPDK processes from running at the same time. So using a FS is
> intentional here.
>

I still think it is stupid:
First of all, why can't someone run two DPDK processes if they each uses a
separate device?
Second, in our case, we don't even use DPDK (at least, I don't think it was
used, right?)! So now DPDK will prevent us from running two scylla
processes on the same host? We often do run (for testing) several scylla
processes on the same host, with different IP addresses, and that *should*
work.


> > Second, yes, @vladzcloudius <https://github.com/vladzcloudius>, the
> > messages could have been clearer - instead of "cannot open" it should
> > have said cannot write, or something of this sort, and avoid this
> > phrase "for rte_mem_config"
> >
> If they fail to even open it why would they report that they failed to
> write into it?
>

You're right technically, but when someone reads a message saying something
like "cannot open config file", he instinctively expects that *he* was
supposed to
create this config file, and scylla was expecting to read it. It's very
unclear that
the code was actually trying to create a new file there.


>
> > (who the heck knows what rte_mem_config is???)
> >
>
> We (ScyllaDB Ltd.) should know. DPDK is not intended to be used directly
> by a "non-aware" users and the messages above are meant to be read by
> the well-aware personnel.


But we are outputting all these messages "not intended to be used directly
by
non-aware users" into the log, where users who know about databases - but
not DPDK, will see them.


> It's up to us to prevent them from being
> printed if don't need them anymore. I don't think it's fair to blame
> DPDK for then thing we haven't done...
>

Ok, then I guess we need to hide most of these messages somehow...
I don't even know how they get printed... Do they use printf()? Do we
implement the logging function and DPDK just calls it?


>
> > Third, most of the messages we see in this log should not have
> > appeared, and the real problem would have stood out more easily. Why
> > do we need to see 40 lines, one for each core detected by DPDK? When
> > we're not even using DPDK? :-)
> >
>
> We do use DPDK. If u see these messages means we use DPDK. In particular
> we use its sharding capability - our seastar's per-shard "engine" is
> running inside DPDK's thread.
>

This doesn't mean we "use" DPDK. It means we decided to make our lives a
tiny bit easier and start the DPDK threads even when DPDK isn't used. As I
said earlier, I think this was a mistake - I think that if we do not plan
to use DPDK
in earnest (to actually drive the network card), we should not have used it
AT ALL
in this run - not start any threads, EAL, RTE, or other TLAs.

>
>
> DPDK maybe calls for rte_panic() in too many places during the
> initialization - that's true.


Can we override what this "rte_panic()" does?
Additionally, if we know that DPDK checks for writability of certain
directory and abort()s if it isn't writable, we can preemptively check it
ourself before starting DPDK, just to be safe.


> We can either cook a patch fixing this
> specific place (and this would not be a single line patch) or we may go
> and make sure in our installation scripts that the permissions are as
> expected. The second would be MUCH easier.
> Or alternatively - we may compile the DPDK out as u suggested and then
> we will have to maintain two distributions - one with DPDK support and
> another without and IMHO this is not justified.
>

The third alternative is not to *compile* out DPDK, but rather to *run* it
out,
i.e., when the "--enable-dpdk" (or whatever) run-time option is not given,
simply don't start any of the DPDK infrastructure. It will still be in the
executable,
but won't run. I know we rejected this option in the past, but I think it's
time
to rethink it.

**Comment by vladzcloudius** [comment]
On 02/11/2016 01:38 PM, nyh wrote:
> On Thu, Feb 11, 2016 at 1:04 PM, vladzcloudius <notifications@github.com>
> wrote:
>
> >
> > On 02/10/2016 11:31 PM, nyh wrote:
> > >
> > > @vladzcloudius <https://github.com/vladzcloudius>, am I correct in
> > > understanding that this ".rte_config" is */not/* a configuration file
> > > which DPDK needs to read, but rather - a file which DPDK tries do
> > /write/?
> > >
> > > First of all, I think it absolutely stupid it uses files at all for
> > > this purpose (it should keep this information in memory).
> > >
> >
> > If memory serves me well DPDK**is also using this file to prevent two
> > /master/ DPDK processes from running at the same time. So using a FS is
> > intentional here.
> >
>
> I still think it is stupid:
> First of all, why can't someone run two DPDK processes if they each uses a
> separate device?

@nyh, I'm afraid you are imaging something much more simplistic when it 
really is when u think about DPDK.
It's not only about managing network devices as it seems to me u think 
it is. It also manages memory and sharding, just like seastar does but a 
bit differently.
Therefore there must be only one master DPDK process that will allocate 
and manage the system resources and there may be client/slave processes 
that may get resources from the master process. Read DPDK manual(s) if u 
want to know more.

> Second, in our case, we don't even use DPDK (at least, I don't think 
> it was
> used, right?)! 

No, u are wrong and I've already mentioned this in my previous post!

> So now DPDK will prevent us from running two scylla
> processes on the same host? We often do run (for testing) several scylla
> processes on the same host, with different IP addresses, and that *should*
> work.

Well, there is this "-no-shconf" option that is already used for an OSv 
variant (you should know that!) that would bypass this locking.
Maybe we should enable it *always*...

>
>
> > > Second, yes, @vladzcloudius <https://github.com/vladzcloudius>, the
> > > messages could have been clearer - instead of "cannot open" it should
> > > have said cannot write, or something of this sort, and avoid this
> > > phrase "for rte_mem_config"
> > >
> > If they fail to even open it why would they report that they failed to
> > write into it?
> >
>
> You're right technically, but when someone reads a message saying 
> something
> like "cannot open config file", he instinctively expects that *he* was
> supposed to
> create this config file, and scylla was expecting to read it. It's very
> unclear that
> the code was actually trying to create a new file there.

I agree, this may be confusing but again - user was not supposed to even 
see this and of course not try to interpret it - it's very technical 
information.
Let's just disable this thing with -no-shconf and close this issue, 
shell we? ;)

>
>
> >
> > > (who the heck knows what rte_mem_config is???)
> > >
> >
> > We (ScyllaDB Ltd.) should know. DPDK is not intended to be used directly
> > by a "non-aware" users and the messages above are meant to be read by
> > the well-aware personnel.
>
>
> But we are outputting all these messages "not intended to be used directly
> by
> non-aware users" into the log, where users who know about databases - but
> not DPDK, will see them.
>
>
> > It's up to us to prevent them from being
> > printed if don't need them anymore. I don't think it's fair to blame
> > DPDK for then thing we haven't done...
> >
>
> Ok, then I guess we need to hide most of these messages somehow...
> I don't even know how they get printed... Do they use printf()? Do we
> implement the logging function and DPDK just calls it?

They print it somehow but I suspect it's some controlled logging. I'll 
take a look at it on Sunday.

>
>
> >
> > > Third, most of the messages we see in this log should not have
> > > appeared, and the real problem would have stood out more easily. Why
> > > do we need to see 40 lines, one for each core detected by DPDK? When
> > > we're not even using DPDK? :-)
> > >
> >
> > We do use DPDK. If u see these messages means we use DPDK. In particular
> > we use its sharding capability - our seastar's per-shard "engine" is
> > running inside DPDK's thread.
> >
>
> This doesn't mean we "use" DPDK. It means we decided to make our lives a
> tiny bit easier and start the DPDK threads even when DPDK isn't used. As I
> said earlier, I think this was a mistake - I think that if we do not plan
> to use DPDK
> in earnest (to actually drive the network card), we should not have 
> used it
> AT ALL
> in this run - not start any threads, EAL, RTE, or other TLAs.

But we DO intend to use it!

>
> >
> >
> > DPDK maybe calls for rte_panic() in too many places during the
> > initialization - that's true.
>
>
> Can we override what this "rte_panic()" does?

Yes. We can code it differently - we have our own DPDK clone in the 
seastar's tree.

**Comment by slivne** [comment]
@raphaelsc the original issue is access to /var/lib/scylla - can we make sure we deal with this - the dpdk part can wait for now

**Comment by slivne** [comment]
@raphaelsc ping ...

**Comment by raphaelsc** [comment]
@slivne"the original issue is access to /var/lib/scylla - can we make sure we deal with this - " At startup, we try to lock data and commitlog directories, and if it fails, we report a message as follow: "ERROR 2016-09-21 11:25:11,196 [shard 0] init - Could not initialize /data/root_data: open: Permission denied" Do you think it's good enough for the user?

**Comment by slivne** [comment]
I am assuming you are compiling without DPDK locally and that is why you are not getting the same message that noam is getting.

If you take the rpm of 1.3/1.2 install on a server and then change the persmission so scylla will not have permission you will get (I am assuming we did not fix that) the bad erro that noam got from the code of DPDK trying to create .rte_config and not able todo that (and then a core) 

this is not OK.

**Comment by raphaelsc** [comment]
@slivne it could be that we will need to interact with dpdk code to get config file path, and check its permission ahead of time (file is outside data and commit dirs BTW), or we can change dpdk code to provide a friendlier message and avoid core dump for this specific case.

**Comment by raphaelsc** [comment]
Patch sent to seastar mailing list: https://groups.google.com/forum/#!topic/seastar-dev/T9sT6VzA5ss

**Comment by raphaelsc** [comment]
@slivne I think this issue can be closed after the following seastar patch: 'make dpdk runtime selection run-time rather than compile-time' (https://github.com/scylladb/seastar/commit/a247dd8b0dd3)

**Comment by tomer-sandler** [comment]
I'm amazed to find that this issue (pretty much the same as what GPS Insight encountered - root:root permissions instead of scylla:scylla on /var/lib/scylla) was reported by Noam at Feb 2016 - see: scylladb/scylla#883

It's marked as resolved and closed long ago, but apparently it's not for
the enterprise version that GPS Insights are using (2018.1).

The original error back in 2016 was:

Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: Cannot open '/var/lib/scylla/.rte_config' for rte_mem_config
It should have been changed per the commit by Raphael to:

Could not initialize /data/root_data: open: Permission denied
@tzach / @shlomi, do you want to re-open it?

**Comment by tomer-sandler** [comment]
I'm amazed to find that this issue (pretty much the same as an issue hit by one of our customers - ```root:root``` permissions instead of ```scylla:scylla``` on ```/var/lib/scylla```) was reported by Noam at Feb 2016 - see: scylladb/scylla#883

It's marked as resolved and closed long ago, but apparently it's not for 2018.1 enterprise version, as a customer hit this issue.

The original error back in 2016 was:

```Feb 08 13:58:28 noam-server-01.localdomain scylla[38851]: Cannot open '/var/lib/scylla/.rte_config' for rte_mem_config```

It should have been changed per the commit by Raphael to:

```Could not initialize /data/root_data: open: Permission denied```

@tzach / @shlomi, do you want to re-open it?
